storagemarket/deal_handler.go

```diff
@@ -6,20 +6,137 @@ import (
 	"fmt"
 	"sync"
 
-	"go.uber.org/atomic"
-
 	"github.com/filecoin-project/boost/storagemarket/types"
+	smtypes "github.com/filecoin-project/boost/storagemarket/types"
+	"github.com/filecoin-project/boost/storagemarket/types/dealcheckpoints"
+	"github.com/google/uuid"
 	"github.com/libp2p/go-eventbus"
 	"github.com/libp2p/go-libp2p-core/event"
-
-	"github.com/google/uuid"
+	"go.uber.org/atomic"
 )
 
+func (p *Provider) mkAndInsertDealHandler(dealUuid uuid.UUID) (*dealHandler, error) {
+	p.dhsMu.Lock()
+	defer p.dhsMu.Unlock()
+	return p.unlkMkAndInsertDealHandler(dealUuid)
+}
+
+func (p *Provider) unlkMkAndInsertDealHandler(dealUuid uuid.UUID) (*dealHandler, error) {
+	dh, ok := p.dhs[dealUuid]
+	if ok {
+		return dh, nil
+	}
+
+	dh, err := newDealHandler(p.ctx, dealUuid)
+	if err != nil {
+		return nil, fmt.Errorf("creating deal handler: %w", err)
+	}
+
+	p.dhs[dealUuid] = dh
+	return dh, nil
+}
+
+func (p *Provider) getDealHandler(id uuid.UUID) *dealHandler {
+	p.dhsMu.RLock()
+	defer p.dhsMu.RUnlock()
+
+	return p.dhs[id]
+}
+
+func (p *Provider) delDealHandler(dealUuid uuid.UUID) {
+	p.dhsMu.Lock()
+	delete(p.dhs, dealUuid)
+	p.dhsMu.Unlock()
+}
+
+// startDealThread sets up a deal handler and wait group monitoring for a deal, then
+// executes the deal in a new go routine
+func (p *Provider) startDealThread(deal *types.ProviderDealState) (bool, error) {
+	p.dhsMu.Lock()
+	defer p.dhsMu.Unlock()
```

Comment:
Instead of having its own lock, the deal thread map now uses the same lock as the deal handlers, for simplicity